ipyparallel/cluster/shellcmd.py

```diff
@@ -0,0 +1,859 @@
+from tempfile import NamedTemporaryFile
+
+
+class Platform(enum.Enum):
+    Unknown = 0
+    Linux = 1
+    Windows = 2
+    MacOS = 3
+
+    @staticmethod
+    def get():
+        import sys
+
+        tmp = sys.platform.lower()
+        if tmp == "win32":
+            return Platform.Windows
+        elif tmp == "linux":
+            return Platform.Linux
+        elif tmp == "darwin":
+            return Platform.MacOS
+        else:
+            raise Exception(f"Unknown platform label '{sys.platform}'")
+
+
+class SimpleLog:
+    def __init__(self, filename):
+        userdir = ""
+        platform = Platform.get()
+        if platform == Platform.Windows:
+            userdir = os.environ["USERPROFILE"]
+        elif platform == Platform.Linux:
+            userdir = os.environ["HOME"] if "HOME" in os.environ else ""
+        self.filename = filename.replace(
+            "${userdir}", userdir
+        )  # replace possible ${userdir} placeholder
+
+    def _output_msg(self, level, msg):
+        with open(self.filename, "a") as f:
+            f.write(f"{level:8} {datetime.now()} {msg}\n")
+
+    def info(self, msg):
+        self._output_msg("info", msg)
+
+    def warning(self, msg):
+        self._output_msg("warning", msg)
+
+    def error(self, msg):
+        self._output_msg("error", msg)
+
+    def debug(self, msg):
+        self._output_msg("debug", msg)
+
+
+class ShellCommandReceive:
+    """
+    Helper class for receiving and performing shell commands in a platform independent form
+
+    All supported shell commands have a cmd_ prefix. When adding new functions make sure that there is an
+    equivalent in the ShellCommandSend class. When a command failed a non-zero exit code will be returned.
+    Hence, the ShellCommandSend class always uses subprocess.check_output for assessing if the command was
+    successful. Some command require information to be returned (cmd_exists, cmd_start, cmd_running) which
+    is written to stdout in the following form: __<key>=<value>__
+    """
+
+    def _log(self, msg):
+        if not self.log:
+            return
+        self.log.info(f"[id={self.ranid}] {msg}")
+
+    def __init__(self, debugging=False, use_breakaway=True, log=None):
+        self.debugging = debugging
+        self.log = None
+        self.platform = Platform.get()
+        self.use_breakaway = use_breakaway
+        assert isinstance(self.use_breakaway, bool)
+        if log:
+            if isinstance(log, str):
+                self.log = SimpleLog(log)
+            else:
+                self.log = log
+        elif "SHELLCMD_LOG" in os.environ:
+            self.log = SimpleLog(os.environ["SHELLCMD_LOG"])
+
+        self.ranid = None
+        if self.log:
+            self.ranid = randint(0, 999)
+            self._log("ShellCommandReceive instance created")
+
+    def __del__(self):
+        if self.log:
+            self._log("ShellCommandReceive instance deleted")
```

Comment:
As mentoined earlier, the log was mainly there to discover the different github windows runner issues and it's not really needed any more.  
So from my point of view, I don't really care if `__del__` is unreliable called or not. I could add a `close` function which doesn't really do anything (log is closed and after each message anyway)

Is there a kind of `close` function in `SSHLauncher` which could close the ssh_sender?

We should find a simple and pragmatic solution to this 'issue'. Please make a suggestion...

Reply:
Hm. No launcher has had ongoing open connections before.

`on_stop` is called when the launcher stops. I think that's a sensible time to close things.

Reply:
If have changed the `SimpleLog` code a bit (see c901aa5828ab8db959a611fed77e0cb9ce77e386). Now it holds a file handle (as you have suggested earlier) and the _shellcmd_ classes have a new `close` function. I have also removed the `__del__` function.

The `ShellCommandReceiveBase.close` has a different semantic as the old `__del__` function; one needs to keep in mind that `ShellCommandReceive` objects are construct (and possibly destructed) for each command that is sent. Hence, it does NOT live as long as the `ShellCommandSend` object (in `SSHLauncher`). So the new `close` function is a signal for the 
receiver that its sender counterpart is going to 'stop'. This would give the receiver a possibility to do some clean up work. But of course we currently don't need that. 

So should we keep the new `close` concept or simply drop it? the calling of the `close` function is done by a `on_stop` handler as suggested (see b6dcf8a31c96780200fe2e4826281ec730267a70)

Reply:
This morging my brain came across another option to clean this up using a context manager. When building the python call for of the [ShellCommandReceive command](https://github.com/ottointhesky/ipyparallel/blob/b6dcf8a31c96780200fe2e4826281ec730267a70/ipyparallel/cluster/shellcmd.py#L322C1-L330C22) we could simply use a context manager there, to restore the original semantic of `__del__`. 
What do you think?

Reply:
`ShellCommandReceive `now uses a [contex manager pattern](https://github.com/ottointhesky/ipyparallel/blob/f0958add192d8b69a29111b5ae37bb297a6189ce/ipyparallel/cluster/shellcmd_receive.py#L299C1-L309C25). I have removed the `on_stop` signal again. So for my point of view, this topic can be closed